File: Civitas - Ruby/pkg/Civitas_Ruby-0.0.1/lib/tablero.rb
# encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class Tablero
    attr_reader :numCasillaCarcel
    attr_reader :porSalida
    attr_reader :casillas
    
    
    def initialize (carcel)
      if carcel >= 1
        @numCasillaCarcel = carcel
      else
        @numCasillaCarcel = 1
      end
      
      @casillas = []
      @casillas.push( Casilla.new("Salida") )
      @porSalida = 0
      @tieneJuez = false      
    end
    
    #Comprobacion de de la integridad del tablero
    def correcto
      return ( (@casillas.length > @numCasillaCarcel) && @tiene_juez )
    end
    
    #Comprobacion de acceso a casilla 
    #Tiene distinto nombre ya que en ruby no se puede tener dos iguales
    def correcto_num numCasilla
      return ( correcto && (numCasilla >= 0) && (numCasilla < @casillas.length) )
    end
    
    #Decrementa el valor de por_salida si es mayor que 0 y lo devuelve
    def get_por_salida
      if @por_salida>0
        @por_salida = @por_salida-1
      end
      return (@por_salida + 1)
    end
    
    #Añade casilla, y si procede, tambien la de carcel
    def añade_casilla casilla
      if @numCasillaCarcel == @casillas.length
        carcel = Casilla.new("Carcel")
        @casillas.push(carcel)
      end
      
      @casillas.push(casilla) 
    
      if @num_casilla_carcel == @casillas.length
        carcel = Casilla.new("Carcel")
        @casillas.push(carcel)
      end
    end
    
    #Añade un juez si no existe
    def añade_juez
      if !@tieneJuez
        juez = Casilla.new("Juez")
        @casillas.push(juez)
        @tiene_juez = true
      end
    end
    
    #Devuelve la casilla solicitada si esta en el rango de casillas existentes
    def get_casilla numCasilla
      if correcto_num(numCasilla)
        return @casillas[numCasilla]
      else
        return nil 
      end
    end
    
    #Calcula la nueva posicion en relacion al origen de la tirada 
    def nueva_posicion(actual, tirada)
      if !correcto
        return -1
      else
        final = actual + tirada

        if final > @casillas.length
          final = final % @casillas.lenght
          @por_salida = @por_salida + 1
        end
        return final
      end
    end
    
    #Emula una tirada para saber cuanto falta para llegar a la posicion deseada
    def calcular_tirada(origen, destino)
      tirada = origen - destino
      if tirada < 0
        tirada = tirada + @casillas.lenght 
      end
      return tirada
    end
    
  end
end

File: Civitas - Java/src/civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
/**
 *
 */
public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    
    Tablero(int indice_casillaCarcel){
        if(indice_casillaCarcel >= 1)
            numCasillaCarcel = indice_casillaCarcel;
        else
            numCasillaCarcel = 1;
        
        casillas = new ArrayList<Casilla>();
        casillas.add(new Casilla("Salida"));
        porSalida = 0;
        tieneJuez = false;
    }
    
    boolean correcto(){
        return ( this.casillas.size() > numCasillaCarcel && this.tieneJuez );
    }
    
    boolean correcto(int numCasilla){
        return ( this.correcto() && (numCasilla < this.casillas.size()) );
    }
    
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        if(porSalida > 0){
            porSalida--;
            return (porSalida+1);
        }
        else
            return 0; //return porSalida;
    }
    
    void añadeCasilla(Casilla casilla){
        if(this.casillas.size() == this.numCasillaCarcel)
            this.casillas.add(new Casilla("Cárcel"));
        this.casillas.add(casilla);
        //NO ME QUEDA CLARO ¿REPETIR AÑADIR CARCEL?¿HAY 2 CARCELES?
        if(this.casillas.size() == this.numCasillaCarcel)
            this.casillas.add(new Casilla("Cárcel"));
    }
    
    void añadeJuez(){
        if(!tieneJuez){
            tieneJuez = true;
            this.casillas.add(new CasillaJuez(numCasillaCarcel, "Juez"));
        }
    }
    
    Casilla getCasilla(int numCasilla){
        if(correcto(numCasilla))
            return casillas.get(numCasilla);
        else
            return null;
    }
    
    int nuevaPosicion(int actual, int tirada){
        int posFinal = actual + tirada;
        if(!this.correcto())
            return -1;
        else{
            if( (posFinal)>this.casillas.size() )
                porSalida++;
                
            return ((posFinal)%this.casillas.size());
        }
    }
    
    int calcularTirada(int origen, int destino){
        int resta = origen - destino;
        if(resta<0)
            return (resta + this.casillas.size());
        else
            return resta;
    }
    
    public void printcasillas(){
        for( Casilla x : casillas  )
            System.out.println("|||||||||||||||||||||||||||||||||||||||||||||||||||||||\n" + x.toString());
               
    }   
    
}